noir-projects/noir-contracts/contracts/amm_contract/src/main.nr

```diff
@@ -0,0 +1,371 @@
+            state,
+            token0_slot_commitment,
+            token1_slot_commitment,
+            liquidity,
+            amount0_min,
+            amount1_min
+        ).enqueue(&mut context);
+    }
+
+    #[public]
+    #[internal]
+    fn _remove_liquidity(
+        // We pass the state as an argument in order to not have to read it from public storage again.
+        state: State,
+        token0_slot_commitment: Field,
+        token1_slot_commitment: Field,
+        liquidity: u64,
+        amount0_min: u64,
+        amount1_min: u64
+    ) {
+        // We don't need any kind of reentrancy guard here because the only way to enter this public function is from
+        // `remove_liquidity` which is private and since public functions cannot call private ones it's impossible to
+        // reenter this function.
+
+        let token0 = Token::at(state.token0);
+        let token1 = Token::at(state.token1);
+        let liquidity_token = Token::at(state.liquidity_token);
+
+        // We get the reserves and the liquidity token total supply.
+        let reserve0 = token0.balance_of_public(context.this_address()).view(&mut context) as u64;
+        let reserve1 = token1.balance_of_public(context.this_address()).view(&mut context) as u64;
+        let total_supply = liquidity_token.total_supply().view(&mut context) as u64;
+
+        // We calculate the amounts of token0 and token1 the user is entitled to based on the amount of liquidity they
+        // are removing.
+        let amount0 = liquidity * reserve0 / total_supply;
+        let amount1 = liquidity * reserve1 / total_supply;
+
+        // We check if the amounts are greater than the minimum amounts the user is willing to accept.
+        assert(amount0 >= amount0_min, "INSUFFICIENT_0_AMOUNT");
+        assert(amount1 >= amount1_min, "INSUFFICIENT_1_AMOUNT");
+
+        // At last we burn the liquidity tokens and transfer the token0 and token1 to the user.
+        liquidity_token.burn_public(context.this_address(), liquidity, 0).call(&mut context);
+        token0.finalize_transfer_to_private(token0_slot_commitment, amount0).call(&mut context);
+        token1.finalize_transfer_to_private(token1_slot_commitment, amount1).call(&mut context);
+    }
+
+    /// Swaps `amount_in` of `token_in` for at least `amount_out_min` of `token_out`. The `from_0_to_1` flag indicates
+    /// whether we are swapping `token0` for `token1` or vice versa. `nonce` can be arbitrary non-zero value and its
+    /// purpose is to isolate authwits to this specific call.
+    #[private]
+    fn swap_exact_tokens_for_tokens(
+        amount_in: u64,
+        amount_out_min: u64,
+        from_0_to_1: bool,
+        nonce: Field
+    ) {
+        let state = storage.state.read_private();
+
+        let (token_address_in, token_address_out) = if from_0_to_1 {
+            (state.token0, state.token1)
+        } else {
+            (state.token1, state.token0)
+        };
```

Comment:
I'd pass `token_in` and `token_out` and do

```rust
assert(token_in != token_out);
assert(token_in == state.token0 || token_in == state.token1);
assert(token_out == state.token0 || token_out == state.token1);
```

Reply:
Addressed in https://github.com/AztecProtocol/aztec-packages/pull/8644/commits/db9146e14c8210b317addd943cf8acd9f1603363